File: algorithms/[713]乘积小于K的子数组/solution.ts
// topics = ["数组", "滑动窗口"]

function numSubarrayProductLessThanK(nums: number[], k: number): number {
  if (k <= 1) {
    return 0;
  }
  const n = nums.length;
  let res = 0;
  let prod = 1;
  let i = 0;
  let j = 0;

  while (j < n) {
    prod *= nums[j];
    while (prod >= k) {
      prod /= nums[i];
      i += 1;
    }
    res += j - i + 1;
    j += 1;
  }

  return res;
}

File: algorithms/[713]乘积小于K的子数组/solution.py
# topics = ["数组", "滑动窗口"]

from typing import List


class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1:
            return 0
        n = len(nums)
        res = 0
        prod = 1
        i = j = 0

        while j < n:
            prod *= nums[j]
            while prod >= k:
                prod //= nums[i]
                i += 1
            # 新增 j - i + 1 个子数组
            res += j - i + 1
            j += 1

        return res
